### ISSUE TITLE
"fatal: morestack on g0" crash

### ISSUE THREAD

**Issue by masiulaniec** [therecipe/qt#399]
Crash:

```
fatal: morestack on g0
SIGTRAP: trace trap
PC=0x4079c02 m=0 sigcode=1
signal arrived during cgo execution

goroutine 1 [running, locked to thread]:
runtime.morestack()
	/usr/local/go/src/runtime/asm_amd64.s:368 +0x22 fp=0x7fff5fb78ec8 sp=0x7fff5fb78ec0

goroutine 17 [syscall, locked to thread]:
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:2197 +0x1

rax    0x17
rbx    0x4bc95a0
rcx    0x407d5a5
rdx    0x0
rdi    0x2
rsi    0x480d62e
rbp    0xc42004d2f0
rsp    0x7fff5fb78ec0
r8     0xc4200001a0
r9     0x2
r10    0x7b100f8
r11    0x206
r12    0x40
r13    0x40
r14    0x1ff0
r15    0x200
rip    0x4079c02
rflags 0x206
cs     0x2b
fs     0x0
gs     0x0
```

Repro:

```
package main

import (
	"os"

	"github.com/therecipe/qt/widgets"
)

func main() {
	widgets.NewQApplication(len(os.Args), os.Args)
	foo := widgets.NewQWidget(nil, 0)
	for i := 0; i < 2000; i++ {
		widgets.NewQLabel(foo, 0)
	}
	w := widgets.NewQMainWindow(nil, 0)
	w.SetCentralWidget(foo)
	w.Show()
	widgets.QApplication_Exec()
}
```

Qt 5.8, Go 1.8.3, macOS 10.12.5, amd64.

**Comment by therecipe** [comment]
Hey

It's running fine for me on windows (32/64) and linux (64).
But crashes on macOS (64) and android (arm) with the same "fatal: morestack on g0" message.
I used Qt 5.8 and Go 1.8.3 for all tests.

It's probably a Go issue: https://github.com/golang/go/issues/20975

You can work around this for now by using qtminimal.
Like this: `qtminimal && go build -tags=minimal ...`

And/or by using a helper moc struct + slot like this: (but then you will need to also use qtmoc)
```go
package main

import (
	"os"

	"github.com/therecipe/qt/core"
	"github.com/therecipe/qt/widgets"
)

type Helper struct {
	core.QObject

	_ func() `slot:"create"`
}

func main() {
	widgets.NewQApplication(len(os.Args), os.Args)
	foo := widgets.NewQWidget(nil, 0)
	helper := NewHelper(nil)
	helper.ConnectCreate(func() {
		for i := 0; i < 2000; i++ {
			println("create ", i)
			widgets.NewQLabel(foo, 0)
		}
	})
	go helper.Create() //<- needs to be invoked in another go routine, so that the call is deferred until the main event loop is running

	w := widgets.NewQMainWindow(nil, 0)
	w.SetCentralWidget(foo)
	w.Show()
	widgets.QApplication_Exec()
}
```

**Comment by masiulaniec** [comment]
Thanks for looking into it.

> It's running fine for me on windows (32/64) and linux (64)

Have you tried bumping the iteration count to a value higher than 2000? I'm not claiming it will crash, I just don't have an easy access to these platforms.

**Comment by therecipe** [comment]
@masiulaniec 
Yeah, on windows 386 is crashes when there are more than 4000 new labels created at once.
On linux amd64 it crashes somewhere between 20-25k new labels.

**Comment by therecipe** [comment]
Okay, I looked a bit deeper into this and the issue is not the same as in golang/go#20975, even though the symptoms are identical. I hoped that raising the stack size like they mentioned there would simply "solve" this in one way or the other (like on linux), but it probably would only conceal the real issue.

The issue here is that the underlying`QWidget::paintEvent` function, will call the `QWidgetPrivate::paintSiblingsRecursive` function, which will then call the `QWidget::paintEvent` function again until all the siblings are painted and the initial "QWidgetPrivate::drawWidget" function can return.
This is normally no problem, but the current implementation of the binding (and/or maybe Go) are not prepared for such a deep recursion between Go and C.

I already had a related issue in mind, but only for performance reasons and therefore pushed it back at the time.

However, I will start to work on this now.
If it really is the same issue, then the chances should also reduce the general memory and cpu usage.

**Comment by therecipe** [comment]
Seems like I hit a dead end.
I tried to remove the unnecessary callback round trips through Go land, but it turned out to neither solve this issue nor to be more efficient.

However, I figured that if show the window (and centralWidget) first before adding the labels, then the labels will be painted one by one rather than being painted in batch.

So this should work:
```go
package main

import (
  "os"

  "github.com/therecipe/qt/widgets"
)

func main() {
  widgets.NewQApplication(len(os.Args), os.Args)
  foo := widgets.NewQWidget(nil, 0)
  w := widgets.NewQMainWindow(nil, 0)
  w.SetCentralWidget(foo)
  w.Show()
  for i := 0; i < 200000; i++ {
    widgets.NewQLabel(foo, 0)
  }
  widgets.QApplication_Exec()
}
```

**Comment by masiulaniec** [comment]
Good finding, thanks.

**Comment by therecipe** [comment]
Don't know why I closed this without further comment earlier.
IIRC this issues fixed itself in Go 1.10 or so, and was somehow an Go issue after all.
